feat: add ranger.finance integration

yongkangc opened the pull request:
# Ranger Integration

https://ranger.finance/


## Changes Made
This PR adds the following changes:
- Implements a full actions for the Ranger API in `packages/plugin-defi/src/ranger`
- Adds all SOR (write/trade) and Data API (read) actions as individual files with Zod schemas and async handlers
- Exports all actions from `actions/index.ts` for easy integration
- Exports all tools (schemas and handlers) from `tools/index.ts` for agentic workflows
- Updates the `README.md` with a full list of features, usage examples, and documentation links

## Implementation Details
- Each Ranger API endpoint is implemented as a separate action file, following the structure and style of the Adrena plugin
- All actions use Zod for input validation and provide minimal, correct async handlers for HTTP calls
- The `actions/index.ts` and `tools/index.ts` files export all available actions and tools for discoverability and agent compatibility
- The `README.md` is updated to reflect all available actions, setup instructions, and usage examples

## Transaction executed by agent 
Example transaction: 
```ts
const quote = await getQuote({
  fee_payer: 'YOUR_PUBLIC_KEY',
  symbol: 'SOL',
  side: 'Long',
  size: 1.0,
  collateral: 10.0,
  size_denomination: 'SOL',
  collateral_denomination: 'USDC',
  adjustment_type: 'Increase',
}, process.env.RANGER_API_KEY);
```

## Additional Notes
- All endpoints are currently set to use the staging Ranger API URLs. Update these to production as needed.
- The structure is modular and ready for further extension or integration with agentic workflows.

## Checklist
- [x] I have tested these changes locally
- [x] I have updated the documentation
- [x] I have added a transaction link
- [x] I have added the prompt used to test it

yongkangc committed:
feat: ranger.finance integration

yongkangc committed:
feat: add new actions and tools for position management and funding rates

yongkangc committed:
feat: refactor API base URLs for ranger actions to use environment variables

yongkangc commented:
@thearyanag would love a review / note here :)

yongkangc committed:
feat: implement action handlers for position management and balance withdrawals in Ranger API

yongkangc committed:
docs: update README to reflect Ranger Finance Plugin for Solana Agent Kit

yongkangc committed:
feat: enhance Ranger API actions with context interfaces and improved type definitions

yongkangc committed:
feat: implement action handlers for funding rates and liquidations in Ranger API with context interfaces

yongkangc commented:
hey @thearyanag bumping up on the review :)

thearyanag commented on packages/plugin-defi/src/ranger/actions/closePosition.ts:
```diff
@@ -0,0 +1,65 @@
+    input: ClosePositionInput,
+    { apiKey }: ClosePositionContext
+  ) => {
+    const response = await fetch(`${RANGER_SOR_API_BASE}/v1/close_position`, {
```
does this don't require the agent's signature?

thearyanag commented on packages/plugin-defi/src/ranger/actions/decreasePosition.ts:
```diff
@@ -0,0 +1,76 @@
+    input: DecreasePositionInput,
+    { apiKey }: DecreasePositionContext
+  ) => {
+    const response = await fetch(
```
same here, does this don't require the agent's signature

yongkangc committed:
feat: update action handlers in Ranger API to use SolanaAgentKit type for improved type safety

yongkangc committed:
fix: apply message pattern

Calls the SOR API endpoint
Decodes the base64 transaction message
Deserializes it to a Solana transaction
Adds a recent blockhash
Signs and sends the transaction
Returns the transaction signature and meta

yongkangc committed:
feat: implement SOR API integration for trading actions in Ranger

- Added API calls for opening, closing, increasing, and decreasing perp trades.
- Implemented balance and collateral withdrawal actions using SOR API.
- Enhanced error handling for API responses and transaction signing process.
- Updated function signatures to include API key and adjusted parameters for improved clarity.

yongkangc committed:
Merge branch 'v2' into v2

yongkangc committed:
docs: enhance README with visual elements and branding for Ranger Finance Plugin

yongkangc committed:
revert

yongkangc committed:
revert

yongkangc committed:
Discard changes to packages/plugin-defi/src/flash/tools/flash_close_trade.ts
